src/NHibernate/Transform/QueryAliasToObjectPropertySetter.cs

```diff
@@ -0,0 +1,48 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+
+namespace NHibernate.Transform
+{
+	[Serializable]
+	public class QueryAliasToObjectPropertySetter
+	{
+		private readonly IEnumerable<FieldInfo> _fields;
+		private readonly IEnumerable<PropertyInfo> _properties;
+
+		private QueryAliasToObjectPropertySetter(FieldInfo[] fields, PropertyInfo[] properties)
+		{
+			_fields = fields;
+			_properties = properties;
+		}
+
+		public static QueryAliasToObjectPropertySetter MakeFor(System.Type objType)
+		{
+			var bindingFlags = BindingFlags.Instance |
+				BindingFlags.Public |
+				BindingFlags.NonPublic |
+				BindingFlags.IgnoreCase;
+			var fields = objType.GetFields(bindingFlags);
+			var properties = objType.GetProperties(bindingFlags);
+
+			return new QueryAliasToObjectPropertySetter(fields, properties);
+		}
+
+		public void SetProperty(string alias, object value, object resultObj)
+		{
+			var property = _properties.SingleOrDefault(prop => string.Equals(prop.Name, alias, StringComparison.OrdinalIgnoreCase));
+			var field = _fields.SingleOrDefault(prop => string.Equals(prop.Name, alias, StringComparison.OrdinalIgnoreCase));
```

Comment:
True, thought this is bad practice to have member name differing only by the case, we have to support it.

The current field/property resolution process does not really support it either, it will take the first that comes, without having any clue whether that is the more sensible one to take or not.

So we can fetch the dictionary with a similar logic, grouping fields/properties by name case insensitively and just putting the first in (while filtering out read only properties by the way, in order to avoid doing that at each field set).

A more sensible algorithm would be to: 1. filter out readonly properties - 2. sort by accessibility then by declaring type inheritance depth (at least, is it inherited or not). 3. take the first (may still have many with same ordering rank).

We could go further by maintaining a case sensitive dictionary too, resolving by it first then fallbacking on the case insensitive one. But that would be added complexity and added behavior discrepancies between DB altering aliases case and others.